src/builders/FulfillmentBuilder.php

```diff
@@ -0,0 +1,263 @@
+
+    /**
+     * @var string $pickupDetailsKey
+     */
+    private string $pickupDetailsKey = 'pickup_details';
+
+    /**
+     * @var string $shipmentDetailsKey
+     */
+    private string $shipmentDetailsKey = 'shipment_details';
+
+    public function __construct()
+    {
+        $this->recipientBuilder = new RecipientBuilder();
+    }
+
+    /**
+     * Checks if the fulfillment details are already set.
+     *
+     * @param mixed $fulfillmentModel
+     *
+     * @throws InvalidSquareOrderException
+     *
+     * @return void
+     */
+    private function checkFulfillmentDetails(mixed $fulfillmentModel): void
+    {
+        // If this fulfillment already has details, throw an error - only one fulfillment details per fulfillment
+        // is currently supported
+        if (!empty($fulfillmentModel->fulfillmentDetails)) {
+            throw new InvalidSquareOrderException('Fulfillment already has details', 500);
+        }
+    }
+
+    /**
+     * Add a fulfillment to the order from model as source.
+     *
+     * @param  Model  $fulfillment
+     * @param  Model  $order
+     * @return Fulfillment|stdClass
+     *
+     * @throws InvalidSquareOrderException
+     */
+    public function createFulfillmentFromModel(Model $fulfillment, Model $order): Fulfillment|stdClass
+    {
+        $fulfillmentObj = new stdClass();
+
+        // Check if the order is present and if it already has this fulfillment
+        // or if fulfillment doesn't have property $id then create new Fulfillment object
+        if (($order && ! $order->hasFulfillment($fulfillment)) && ! Arr::has($fulfillment->toArray(), 'id')) {
+            $tempFulfillment = new Fulfillment($fulfillment->toArray());
+        } else {
+            $tempFulfillment = Fulfillment::find($fulfillment->id);
+        }
+
+        // Validate that the type matches the details
+        // Due to the nature of the relationship between Fulfillment and FulfillmentDetails, the fulfillment details
+        // should already be a model associated with the fulfillment at this point
+        if (
+            $fulfillment->type == Constants::FULFILLMENT_TYPE_DELIVERY
+            && ! $fulfillment->fulfillmentDetails instanceof DeliveryDetails
+        ) {
+            throw new InvalidSquareOrderException('Fulfillment type does not match details', 500);
+        } elseif (
+            $fulfillment->type == Constants::FULFILLMENT_TYPE_PICKUP
+            && ! $fulfillment->fulfillmentDetails instanceof PickupDetails
+        ) {
+            throw new InvalidSquareOrderException('Fulfillment type does not match details', 500);
+        } elseif (
+            $fulfillment->type == Constants::FULFILLMENT_TYPE_SHIPMENT
+            && ! $fulfillment->fulfillmentDetails instanceof ShipmentDetails
+        ) {
+            throw new InvalidSquareOrderException('Fulfillment type does not match details', 500);
+        }
+
+        $fulfillmentObj = $tempFulfillment;
```

Comment:
Yeah, that's a good call.  I was copying the format here from the product builder. Not something I think needs changing in this PR, but worth noting that there may be an unnecessary reassignment over here:
https://github.com/NikolaGavric94/laravel-square/blob/master/src/builders/ProductBuilder.php#L167

Reply:
Updated this to use the tempFulfillment throughout the function instead of reassigning here:
https://github.com/Mrkbingham/laravel-square/commit/c295919974780205a92f7116dcf706b628ee707d

Reply:
Minor tweak on this one as the original refactor here wasn't correct: https://github.com/Mrkbingham/laravel-square/commit/ddb8dc6b4e5c172dc6165a00efb169cd6806ed2f